pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,289 @@
+
+	minio "github.com/minio/minio-go"
+	"github.com/minio/minio-go/pkg/credentials"
+	"github.com/rancher/kontainer-engine/drivers/rke"
+	"github.com/rancher/kontainer-engine/service"
+	"github.com/rancher/rancher/pkg/controllers/management/clusterprovisioner"
+	"github.com/rancher/rancher/pkg/rkedialerfactory"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	clusterBackupCheckInterval = 5 * time.Minute
+)
+
+type Controller struct {
+	ctx                   context.Context
+	clusterClient         v3.ClusterInterface
+	clusterLister         v3.ClusterLister
+	backupClient          v3.EtcdBackupInterface
+	backupLister          v3.EtcdBackupLister
+	backupDriver          service.EngineService
+	secretsClient         v1.SecretInterface
+	KontainerDriverLister v3.KontainerDriverLister
+}
+
+func Register(ctx context.Context, management *config.ManagementContext) {
+	c := &Controller{
+		ctx:                   ctx,
+		clusterClient:         management.Management.Clusters(""),
+		clusterLister:         management.Management.Clusters("").Controller().Lister(),
+		backupClient:          management.Management.EtcdBackups(""),
+		backupLister:          management.Management.EtcdBackups("").Controller().Lister(),
+		backupDriver:          service.NewEngineService(clusterprovisioner.NewPersistentStore(management.Core.Namespaces(""), management.Core)),
+		secretsClient:         management.Core.Secrets(""),
+		KontainerDriverLister: management.Management.KontainerDrivers("").Controller().Lister(),
+	}
+
+	local := &rkedialerfactory.RKEDialerFactory{
+		Factory: management.Dialer,
+	}
+	docker := &rkedialerfactory.RKEDialerFactory{
+		Factory: management.Dialer,
+		Docker:  true,
+	}
+	driver := service.Drivers[service.RancherKubernetesEngineDriverName]
+	rkeDriver := driver.(*rke.Driver)
+	rkeDriver.DockerDialer = docker.Build
+	rkeDriver.LocalDialer = local.Build
+	rkeDriver.WrapTransportFactory = docker.WrapTransport
+
+	c.backupClient.AddLifecycle(ctx, "etcdbackup-controller", c)
+	go c.clusterBackupSync(ctx, clusterBackupCheckInterval)
+}
+
+func (c *Controller) Create(b *v3.EtcdBackup) (runtime.Object, error) {
+	cluster, err := c.clusterClient.Get(b.ClusterID, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig == nil {
+		return nil, fmt.Errorf("[etcd-backup] cluster doesn't have a backup config")
+	}
+
+	if !v3.BackupConditionCompleted.IsTrue(b) {
+		b.Filename = getBackupFilename(b.Name, cluster)
+		b.BackupConfig = *cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig
+		v3.BackupConditionCompleted.True(b)
```

Comment:
I thought this was BackupConditionCreated, and Completed would be only after ETCDSave completes?

Reply:
yes

Reply:
Looks like a typo that needs to be updated @moelsayed ; from completed to created

Reply:
fix, it was a typo.